app/models/highlight_suggestion.rb

```diff
@@ -0,0 +1,68 @@
+class HighlightSuggestion < ApplicationRecord
+  belongs_to :run
+
+  # hunt looks on Twitch for past broadcasts whose timestamps imply that it contains the run's PB. If found, the
+  # highlight suggestion is saved to the database. If not, it is not.
+  def hunt!
+    return if run.user.nil? || run.user.twitch.nil?
+
+    pb = run.histories.find_by(
+      realtime_duration_ms: run.duration_ms(Run::REAL),
+      gametime_duration_ms: run.duration_ms(Run::GAME)
+    )
+    return if pb.nil?
+
+    Twitch::Videos.recent(run.user.twitch.twitch_id).each do |video|
+      match = /^((\d+)h)?((\d+)m)?((\d+)s)?$/.match(video['duration'])
+      hours   = match[2].to_i.hours
+      minutes = match[4].to_i.minutes
+      seconds = match[6].to_i.seconds
+
+      video_start = DateTime.parse(video['created_at'])
+      video_end   = video_start + hours + minutes + seconds
+
+      next unless video['type'] == 'archive'
+
+      if video_start - 30.seconds < pb.started_at && video_end + 30.seconds > pb.ended_at
+        update(
+          url: URI.parse("https://www.twitch.tv/#{run.user.twitch.name}/manager/highlighter/#{video['id']}").tap do |uri|
+            uri.query = {
+              start: (pb.started_at - video_start - 10.seconds).to_i,
+              end: ((pb.started_at - video_start) + (pb.duration_ms(Run::REAL) / 1000) + 10.seconds).to_i,
+              title: "PB: #{run.game} #{run.category} in #{format_ms(run.duration_ms(run.default_timing))}"
+            }.to_query
+          end
+        )
+        delay(run_at: 60.days.from_now).destroy # 60 days is life of archives for Partners / Twitch Prime members
+        return
+      end
+    end
+  end
+
+  private
+
+  # format_ms accepts a number of milliseconds and returns a time like "HH:MM:SS". If precise is true, it returns a time
+  # like "HH:MM:SS.mmm" instead.
+  def format_ms(milliseconds, precise: false)
+    return '-' if milliseconds.nil?
+    time = explode_ms(milliseconds)
+
+    return format('%02d:%02d:%02d.%03d', time[:h], time[:m], time[:s], time[:ms]) if precise
+    format('%02d:%02d:%02d', time[:h], time[:m], time[:s])
+  end
+
+  # explode_ms returns a hash with the components of the given duration separated out. The returned hash is guaranteed
+  # to be ordered by component size descending (hours before minutes, etc.).
+  def explode_ms(total_milliseconds)
```

Comment:
Same as above.

Reply:
I bet I typed in my comment and forgot to actually hit the comment button.

I was saying is there a way to make this and the `format_ms` functions in such a way that the model and views would be able to access them, like a class method on the `Run` model or something?  It seems hacky to have the methods defined the exact same way in 2 different places because it will probably come back to bite someone at some point.

Reply:
Yeah I struggled with this. Class methods on `Run` is probably good enough. It just feels like `format_ms` is 100% a view-oriented function, so it should go in a helper. But here we are in a model performing the exact same thing just so we can pass this string to Twitch.

Putting it on a model would work, but it would be admitting defeat and having the source of truth for a view-related function be in a model :( but I think you're right overall, that's not as big a deal as duplicating the function.

 If only ActiveSupport::Duration supported sub-seconds. Maybe long-term we should make our own duration model or gem.